CIP-0031/README.md

```diff
@@ -0,0 +1,206 @@
+
+### Why UTXOs?
+
+There are various approaches to augmenting Cardano with the ability to store and retrieve data more easily.
+All of these require some kind of resource-like system for the _storage_ that is used for data.
+The main argument in favour of the reference input approach is that it reuses our existing resource-like system: UTXOs.
+
+As we've seen above, this allows us to satisfy most of our requirements for free.
+Any other approach would need new solutions to these problems.
+
+As a brief example, suppose we wanted to instead implement data storage as an on-chain, hash-indexed data store (a proposal we considered).
+
+Then we would need to answer a lot of questions:
+- How is the data accessed and retrieved? Are these stateful operations? Does this make transaction ordering more significant, threatening determinism?
+- How is the storage usage controlled? Do people pay for storing data? Do they pay for a fixed period, or perpetually? How is data retired?
+- How is the storage going to be implemented? How will this affect node memory usage?
+- How are scripts going to access the data in the store?
+- How will tools interact with and visualize the store and changes to it?
+
+### How should we present the information to scripts?
+
+Scripts definitely need to see reference inputs.
+But we have at least two options for how we represent this in the script context: put the reference inputs in their own field; or include them with the other inputs, but tag them appropriately.
+
+Keeping them separate seems wise given the potential for confusing reference inputs for normal inputs.
+That would be quite a dangerous programming error, as it might lead a script to believe that e.g. an output had been spent when in fact it had only been referenced.
+
+### Accessing the datums of reference inputs
+
+Currently this proposal has somewhat limited utility, because datums in outputs are just hashes, and the spending party is required to find and provide the preimage themselves.
+Therefore the CIP-32 proposal for inline datums is complementary, as it allow UTXOs to carry the data itself rather than a hash.
+
+In the mean time (or if CIP-32 is not implemented), we still want users to be _able_ to provide the datums corresponding to datum hashes in reference outputs, so that at least it is possible to reference the data, albeit clumsily.
+The easiest solution here is to reuse the existing mechanism for providing pre-images of datum hashes which appear in outputs.
+
+Providing datum hash pre-images remains optional, since there are reasons to use reference inputs even without looking at the datum, e.g. to look at the value in an output.
+
+### Accessing the value in outputs
+
+The motivation of this proposal mainly requires looking at the _datums_ of outputs.
+But reference inputs allow us to do more: they let us look at the _value_ in an output.
+
+This is a bonus feature, in that we don't think it's critical, but it is still useful.
```

Comment:
It's definitely critical. Information without any proof of authenticity is not worth much.

Reply:
This section does not concern proof of authenticity. The OP was suggesting the assets, Ada and non-native tokens, stored at UTxO could also be accessed. 

The motivating use case, access state in a datum, can be addressed without the ability to access the value on the UTxO. I agree with the OP. This is a bonus feature.

Reply:
I'm not sure I understand what you mean? This CIP in isolation, without considering the other ones, is completely useless if you can not see the value contained in the UTXO.

Reply:
@L-as I think this comes down to a confusing over-loading of the word "Value".

In one sense, "Value" could mean "The data which hashes to the datum hash", in which case the previous section addresses how this data is made available.

In another sense, "Value" could mean "The ADA and cardano-native-tokens stored in the UTXO".  So, beyond just seeing that the referenced datum has X hash and Y data, you could also see that it contains 133 ADA and 7 HOSKY tokens.

I believe the "bonus feature, in that we don't think it's critical" here refers to this second meaning.

Reply:
I'm not sure whether we're disagreeing, but actually I was wrong here, this *is* very important. The reason for that is that the main way we have of identifying *specific* UTXOs is via their (value) contents, in particular we usually use tokens for this. So if you care about looking at particular UTXOs this is very important. I'll amend.

Reply:
I've attempted to clarify in the text.

Reply:
I agree that this is critical. NFTs are a major way to pinpoint UTXOs of interest.

Without accessing a particular NFT inside a reference input's "Value", I think it would be hard to prove to a script that you're providing it with authentic data, because the adversary can always create UTXOs with arbitrary data under the script address.